core/src/main/java/io/questdb/griffin/engine/join/AsOfJoinNoKeyFastRecordCursorFactory.java

```diff
@@ -0,0 +1,369 @@
+        if (slaveRec.getTimestamp(slaveTimestampIndex) > masterTimestamp) {
+            return lo - 1;
+        }
+        return lo;
+    }
+
+    @Override
+    protected void _close() {
+        ((JoinRecordMetadata) getMetadata()).close();
+        masterFactory.close();
+        slaveFactory.close();
+    }
+
+    private static class AsOfJoinFastRecordCursor implements NoRandomAccessRecordCursor {
+        private final int columnSplit;
+        private final int lookahead;
+        private final int masterTimestampIndex;
+        private final OuterJoinRecord record;
+        private final int slaveTimestampIndex;
+        private boolean isMasterHasNextPending;
+        // stands for forward/backward scan through slave's time frames
+        private boolean isSlaveForwardScan;
+        private boolean isSlaveOpenPending;
+        private RecordCursor masterCursor;
+        private boolean masterHasNext;
+        private Record masterRecord;
+        private TimeFrameRecordCursor slaveCursor;
+        private int slaveFrameIndex = -1;
+        private long slaveFrameRow = Long.MIN_VALUE;
+        private Record slaveRecA;
+        private Record slaveRecB;
+        private long slaveTimestamp = Long.MIN_VALUE;
+
+        public AsOfJoinFastRecordCursor(
+                int columnSplit,
+                Record nullRecord,
+                int masterTimestampIndex,
+                int slaveTimestampIndex,
+                int lookahead
+        ) {
+            this.columnSplit = columnSplit;
+            this.record = new OuterJoinRecord(columnSplit, nullRecord);
+            this.masterTimestampIndex = masterTimestampIndex;
+            this.slaveTimestampIndex = slaveTimestampIndex;
+            this.lookahead = lookahead;
+        }
+
+        @Override
+        public void calculateSize(SqlExecutionCircuitBreaker circuitBreaker, Counter counter) {
+            masterCursor.calculateSize(circuitBreaker, counter);
+        }
+
+        @Override
+        public void close() {
+            masterCursor = Misc.free(masterCursor);
+            slaveCursor = Misc.free(slaveCursor);
+        }
+
+        @Override
+        public Record getRecord() {
+            return record;
+        }
+
+        @Override
+        public SymbolTable getSymbolTable(int columnIndex) {
+            if (columnIndex < columnSplit) {
+                return masterCursor.getSymbolTable(columnIndex);
+            }
+            return slaveCursor.getSymbolTable(columnIndex - columnSplit);
+        }
+
+        @Override
+        public boolean hasNext() {
+            if (isMasterHasNextPending) {
+                masterHasNext = masterCursor.hasNext();
+                isMasterHasNextPending = false;
+            }
+            if (masterHasNext) {
+                final long masterTimestamp = masterRecord.getTimestamp(masterTimestampIndex);
+                if (masterTimestamp < slaveTimestamp) {
```

Comment:
`slaveTimestamp` has the value of the timestamp greater than the matching slaveTimestamp, perhaps name should be more descriptive.

Reply:
Renamed to `lookaheadTimestamp` in b89b1e7645868bddefd4d13d3ebce4f6a9d8ea28